interfaces/builtin/adb.go

```diff
@@ -0,0 +1,219 @@
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/snap"
+)
+
+const adbSummary = `allows operating as Android Debug Bridge service`
+
+const adbBaseDeclarationSlots = `
+  adb:
+    allow-installation:
+      slot-snap-type:
+        - app
+    deny-auto-connection: true
+`
+
+var adbVendorList = map[int]string{
+	0x03f0: "HP",
+	0x03fc: "ECS",
+	0x0408: "QUANTA",
+	0x0409: "NEC",
+	0x0414: "GIGABYTE",
+	0x0451: "TI",
+	0x0471: "PHILPS",
+	0x0482: "KYOCERA",
+	0x0489: "FOXCONN",
+	0x04b7: "COMPAL",
+	0x04c5: "FUJITSU",
+	0x04da: "PMC-SIERRA",
+	0x04dd: "SHARP",
+	0x04e8: "SAMSUNG",
+	0x0502: "ACER",
+	0x0531: "WACOM",
+	0x054c: "SONY",
+	0x05c6: "Qualcomm",
+	0x067e: "INTERMEC",
+	0x091e: "GARMIN-ASUS",
+	0x0930: "TOSHIBA",
+	0x0955: "NVIDIA",
+	0x0b05: "ASUS",
+	0x0bb4: "HTC",
+	0x0c2e: "HONEYWELL",
+	0x0db0: "MSI",
+	0x0e79: "ARCHOS",
+	0x0e8d: "MTK",
+	0x0f1c: "FUNAI",
+	0x0fce: "SONY ERICSSON",
+	0x1004: "LGE",
+	0x109b: "HISENSE",
+	0x10a9: "PANTECH",
+	0x1219: "COMPALCOMM",
+	0x12d1: "HUAWEI",
+	0x1662: "POSITIVO",
+	0x16d5: "ANYDATA",
+	0x17ef: "LENOVO",
+	0x18d1: "GOOGLE",
+	0x1949: "LAB126",
+	0x19a5: "HARRIS",
+	0x19d2: "ZTE",
+	0x1b8e: "AMLOGIC",
+	0x1bbb: "T_AND_A",
+	0x1d09: "TECHFAITH",
+	0x1d45: "QISDA",
+	0x1d4d: "PEGATRON",
+	0x1d91: "BYD",
+	0x1e85: "GIGASET",
+	0x1ebf: "YULONG_COOLPAD",
+	0x1f3a: "ALLWINNER",
+	0x1f53: "SK TELESYS",
+	0x2006: "LENOVOMOBILE",
+	0x201e: "HAIER",
+	0x2080: "NOOK",
+	0x2116: "KT TECH",
+	0x2207: "ROCKCHIP",
+	0x2237: "KOBO",
+	0x2257: "OTGV",
+	0x22b8: "MOTOROLA",
+	0x22d9: "OPPO",
+	0x2314: "INQ_MOBILE",
+	0x2340: "TELEEPOCH",
+	0x2420: "IRIVER",
+	0x24e3: "K-TOUCH",
+	0x25e3: "LUMIGON",
+	0x2717: "XIAOMI",
+	0x271d: "GIONEE",
+	0x2836: "OUYA",
+	0x2916: "YOTADEVICES",
+	0x297f: "EMERGING_TECH",
+	0x29a9: "SMARTISAN",
+	0x29e4: "PRESTIGIO",
+	0x2a45: "MEIZU",
+	0x2a47: "BQ",
+	0x2a49: "UNOWHY",
+	0x2ae5: "FAIRPHONE",
+	0x413c: "DELL",
+	0x8087: "INTEL", // https://twitter.com/zygoon/status/1032233406564913152
+	0xE040: "VIZIO",
+}
+
+var adbPermanentSlotAppArmor = `
+# Allow adb (server) to access all usb devices and rely on the device cgroup for mediation.
+/dev/bus/usb/[0-9][0-9][0-9]/[0-9][0-9][0-9] rw,
+
+# Allow access to udev meta-data about character devices with major number 189
+# as per https://www.kernel.org/doc/Documentation/admin-guide/devices.txt
+# those describe "USB serial converters - alternate devices".
+/run/udev/data/c189:* r,
+
+# Allow reading the serial number of all the USB devices.
+/sys/devices/**/usb*/*/serial r,
+
+# Allow adb (server) to use TCP/IP for localhost IPC.
+# TODO: once fine-grained network mediation is possible, constrain this to localhost.
+network inet stream,
```

Comment:
It seems a bit strange that this interface grants access to all USB devices in the machine to the snap, and is unconditional (permanent) feels problematic. It's also a bit awkward that the slot would fit in the snap that runs the daemon. In general my feeling is that snaps will end up embedding adb instead of using it from a third-party. Who's going to be using that interface, so we can have some more insight into it?

Reply:
@niemeyer - We could make these more specific if we interrogated the system and only added the correct paths (currently this is the purpose of the device cgroup).

You are right to call out that these broad glob rules shouldn't be in PermanentSlotAppArmor; indeed, adb server shouldn't crash in this situation (meaning there is little argument for them being permanent) since it is possible to start the server without any connected devices.

What we want is the permanent *udev* for the slot to install the (slightly modified) upstream udev rules. Then we use a connected apparmor slot to add the wide apparmor glob rules along with the connected udev slot to add the devices to the slot's cgroup. In this manner, the adb server is allowed to only access the tagged devices via normal interface connections. Sorry for not mentioning this sooner.

With this in mind, do we need the connected udev plug? Ie, does the adb command need access to the devices, or does it just need access to the server?

Reply:
@niemeyer - "What we want is the permanent *udev* for the slot to install the (slightly modified) upstream udev rules."

This is why the implementation is the way it is. adb requires udev rules to be installed on the system for users to access devices. I did not want snapd to unconditionally add these rules to every system where snapd is running (largely due to https://github.com/snapcore/snapd/pull/5170#discussion_r212463306), but to instead only add them if a slot implementation of adb is installed on the system. IME, it is reasonable for a slot implementation of the adb server to be on the system and for another snap to plugs into it. I also don't think it is unreasonable for a part to ship the server plus the client and to slots the interface.

Reply:
Thank you for raising this. I will do some testing and adjust the interface based on your collective feedback.